src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/CallArranger.java

```diff
@@ -0,0 +1,466 @@
+
+            // Allocate registers.
+            int fpRegCnt = count;
+            // Rest will get put into a struct. Compute number of 64 bit slots.
+            int structSlots = 0;
+            boolean needOverlapping = false; // See "no partial DW rule" below.
+
+            int availableFpRegs = MAX_FLOAT_REGISTER_ARGUMENTS - nRegs[StorageType.FLOAT];
+            if (count > availableFpRegs) {
+                fpRegCnt = availableFpRegs;
+                int remainingElements = count - availableFpRegs;
+                if (elementCarrier == float.class) {
+                    if ((fpRegCnt & 1) != 0) {
+                        needOverlapping = true;
+                        remainingElements--; // After overlapped one.
+                    }
+                    structSlots = (remainingElements + 1) / 2;
+                } else {
+                    structSlots = remainingElements;
+                }
+            }
+
+            VMStorage[] source = (forArguments ? C.inputStorage : C.outputStorage)[StorageType.FLOAT];
+            VMStorage[] result = new VMStorage[fpRegCnt + structSlots];
+            if (elementCarrier == float.class) {
+                // Mark elements as single precision (32 bit).
+                for (int i = 0; i < fpRegCnt; i++) {
+                    VMStorage sourceReg = source[nRegs[StorageType.FLOAT] + i];
+                    result[i] = new VMStorage(StorageType.FLOAT, PPC64Architecture.REG32_MASK,
+                                              sourceReg.indexOrOffset());
+                }
+            } else {
+                for (int i = 0; i < fpRegCnt; i++) {
+                    result[i] = source[nRegs[StorageType.FLOAT] + i];
+                }
+            }
+
+            nRegs[StorageType.FLOAT] += fpRegCnt;
+            // Reserve GP regs and stack slots for the packed HFA (when using single precision).
+            int gpRegCnt = (elementCarrier == float.class) ? ((fpRegCnt + 1) / 2)
+                                                           : fpRegCnt;
+            nRegs[StorageType.INTEGER] += gpRegCnt;
+            stackAlloc(fpRegCnt * elementSize, STACK_SLOT_SIZE);
+
+            if (needOverlapping) {
+                // "no partial DW rule": Mark first stack slot to get filled.
+                // Note: Can only happen with forArguments = true.
+                VMStorage overlappingReg = null;
+                if (nRegs[StorageType.INTEGER] <= MAX_REGISTER_ARGUMENTS) {
+                    VMStorage allocatedGpReg = C.inputStorage[StorageType.INTEGER][nRegs[StorageType.INTEGER] - 1];
+                    overlappingReg = new VMStorage(StorageType.INTEGER_AND_FLOAT,
+                                                   PPC64Architecture.REG64_MASK, allocatedGpReg.indexOrOffset());
+                } else {
+                    overlappingReg = new VMStorage(StorageType.STACK_AND_FLOAT,
+                                                   (short) STACK_SLOT_SIZE, (int) stackOffset - 4);
```

Comment:
Thanks! I need to find extra time for this. Sounds like a good idea and I may be able to get rid of some nasty code.

Reply:
Done by https://github.com/openjdk/jdk/pull/12708/commits/98e242c24c07ea977b7709b9f8d0c10ce87e84c0 (using a record instead of a `VMStorage[][]` because I think this is better readable). Note that it's a bit more complicated. I couldn't use your `dup` trick, because I need to put the value into a GP reg and one half of it to a FP reg. The Panama code doesn't support that (IllegalArgumentException: Invalid operand type: interface java.lang.foreign.MemorySegment. float expected).

Reply:
Thanks, your solution looks good.